### ISSUE TITLE
apt-fast not "fast" on Windows Subsystem for Linux

### ISSUE THREAD

**Issue by chrstphrchvz** [ilikenwf/apt-fast#142]
Using apt-fast 1.9.4-2 on Debian 10 testing on Windows 10 x64 version 1803.

apt-fast runs very slowly during the `Working... this may take a while` phase and when printing the list of packages. E.g. with `_APTMGR=apt`, running `apt-fast full-upgrade` takes 36 seconds before the list of packages begins printing and another 9 seconds for it to finish printing all 151 packages. Running apt or apt-get command directly does not have this issue, and on a normal Linux system apt-fast is able to do this step almost as quickly as the underlying `_APTMGR` command.

This appears to be caused by /tmp (where `DLLIST=/tmp/apt-fast.list` resides by default) not being RAM-based on WSL. In fact all tmpfs mounts are disk-only currently, and apparently without write buffering (as it saturates the SSD's write activity at about 3MB/s). I haven't yet found where the status of WSL implementing RAM-base tmpfs can be tracked.

I'm not expecting apt-fast to workaround this, nor do I have a good idea what one would be. E.g. even if a named FIFO were to provide better performance, it's not as simple as making `DLLIST` a named FIFO since apt-fast (1) reads `DLLIST` only after it finishes writing it, and (2) reads `DLLIST` more than once.

**Comment by Lasall** [comment]
A workaround could be to introduce another option that holds the dllist in
ram for apt-fast itself, only aria would need to read that file again (or
is there a way we can pass everything to aria skipping the file creation?)

**Comment by Lasall** [comment]
If you drop the DLLIST file size check at https://github.com/ilikenwf/apt-fast/blob/00fe0f072bcdd52a6067dad0a8f26214e13c29cd/apt-fast#L541 how much does it save on your testing environment?

**Comment by chrstphrchvz** [comment]
> If you drop the DLLIST file size check at https://github.com/ilikenwf/apt-fast/blob/00fe0f072bcdd52a6067dad0a8f26214e13c29cd/apt-fast#L541 how much does it save on your testing environment?

Without `[ "$(wc -l "$DLLIST" | cut -d' ' -f1)" -gt 1 ]` it still takes about as long, probably because it didn't reduce the writing.

I think my realtime antivirus (McAfee LifeSafe) might be partly to blame, since disabling it prevents it from writing to one of its logfiles and no longer causes write operations to saturate or CPU usage to increase. There is still some occasional spike in writes (about 300kB/s)--but I'm not sure these are even from Linux (WSL doesn't support iotop). The time taken is reduced significantly to about 29s when it's disabled, but that's still unusually long. I'm less sure what the culprit is now.

I'm not experienced with bash, but here's what I attempted so far to use a string `my_dllist` instead of a file for `DLLIST` (it does not appear to reduce the time taken any further):
```diff
diff --git a/apt-fast b/apt-fast
index ba5c28c..2e0a71b 100755
--- a/apt-fast
+++ b/apt-fast
@@ -69,7 +69,6 @@ TMP_DOWNLOADBEFORE="${DOWNLOADBEFORE-${TMP_RANDOM}}"
 TMP__APTMGR="${_APTMGR-${TMP_RANDOM}}"
 TMP_APTCACHE="${APTCACHE-${TMP_RANDOM}}"
 TMP_DLDIR="${DLDIR-${TMP_RANDOM}}"
-TMP_DLLIST="${DLLIST-${TMP_RANDOM}}"
 TMP_LISTDIR="${LISTDIR-${TMP_RANDOM}}"
 TMP__MAXNUM="${MAXNUM-${TMP_RANDOM}}"
 TMP__MAXCONPERSRV="${MAXCONPERSRV-${TMP_RANDOM}}"
@@ -130,7 +129,6 @@ else
 fi
 # Currently not needed.
 eval "$(apt-config shell LISTDIR Dir::State::lists/d)"
-DLLIST="/tmp/apt-fast.list"
 _MAXNUM=5
 _MAXCONPERSRV=10
 _SPLITCON=8
@@ -158,7 +156,7 @@ DOWNLOADBEFORE=
 VERBOSE_OUTPUT=

 # Download command.
-_DOWNLOADER='aria2c --no-conf -c -j ${_MAXNUM} -x ${_MAXCONPERSRV} -s ${_SPLITCON} -i ${DLLIST} --min-split-size=${_MINSPLITSZ} --stream-piece-selector=${_PIECEALGO} --connect-timeout=600 --timeout=600 -m0'
+_DOWNLOADER='aria2c --no-conf -c -j ${_MAXNUM} -x ${_MAXCONPERSRV} -s ${_SPLITCON} -i <(echo "$my_dllist") --min-split-size=${_MINSPLITSZ} --stream-piece-selector=${_PIECEALGO} --connect-timeout=600 --timeout=600 -m0'

 # Load config file.
 CONFFILE="/etc/apt-fast.conf"
@@ -178,7 +176,6 @@ https_proxy=
 [ "$TMP__APTMGR" = "$TMP_RANDOM" ] || _APTMGR="$TMP__APTMGR"
 [ "$TMP_APTCACHE" = "$TMP_RANDOM" ] || APTCACHE="$TMP_APTCACHE"
 [ "$TMP_DLDIR" = "$TMP_RANDOM" ] || DLDIR="$TMP_DLDIR"
-[ "$TMP_DLLIST" = "$TMP_RANDOM" ] || DLLIST="$TMP_DLLIST"
 [ "$TMP_LISTDIR" = "$TMP_RANDOM" ] || LISTDIR="$TMP_LISTDIR"
 [ "$TMP__MAXNUM" = "$TMP_RANDOM" ] || _MAXNUM="$TMP__MAXNUM"
 [ "$TMP__MAXCONPERSRV" = "$TMP_RANDOM" ] || _MAXCONPERSRV="$TMP__MAXCONPERSRV"
@@ -239,17 +236,7 @@ _remove_lock()
 CLEANUP_STATE=0
 cleanup_dllist()
 {
-  if [ -f "$DLLIST" ]
-  then
-    if ! mv -- "$DLLIST{,.old}" 2>/dev/null
-    then
-      if ! rm -f -- "$DLLIST" 2>/dev/null
-      then
-        msg "Could not clean up download list file." "warning"
-        CLEANUP_STATE=1
-      fi
-    fi
-  fi
+    :
 }

 cleanup_aptfast()
@@ -299,23 +286,10 @@ DOWNLOAD_DISPLAY=
 DOWNLOAD_SIZE=0
 # Get the package URLs.
 get_uris(){
-  if [ ! -d "$(dirname "$DLLIST")" ]
-  then
-    if ! mkdir -p -- "$(dirname "$DLLIST")"
-    then
-      msg "Could not create download file directory." "warning"
-      exit 1
-    fi
-  elif [ -f "$DLLIST" ]; then
-    if ! rm -f -- "$DLLIST" 2>/dev/null && ! touch -- "$DLLIST" 2>/dev/null
-    then
-      msg "Unable to write to download file. Try restarting with root permissions or run 'apt-fast clean' first." "warning"
-      exit 1
-    fi
-  fi

   # Add header to overwrite file.
-  echo "# apt-fast mirror list: $(date)" > "$DLLIST"
+  #echo "# apt-fast mirror list: $(date)" > "$DLLIST"
+  printf -v my_dllist "apt-fast mirror list: %s\n" "$(date)"
   #NOTE: aptitude doesn't have this functionality, so we use apt-get to get
   #      package URIs.
   case "$_APTMGR" in
@@ -403,24 +377,22 @@ get_uris(){
       hash_algo=
     fi

-    {
-      get_mirrors "$uri"
-      #echo " dir=$DLDIR"
+      my_dllist+="$(get_mirrors "$uri")"$'\n'
+      #my_dllist+=" dir=$DLDIR"$'\n'
       if [ -n "$hash_algo" ]; then
-        echo " checksum=$hash_algo=$checksum"
+        my_dllist+=" checksum=$hash_algo=$checksum"$'\n'
       fi
-      echo " out=$filename"
-    } >> "$DLLIST"
+      my_dllist+=" out=$filename"$'\n'
+
   done <<<"$(echo "$uris_full" | grep -E "^'(http(s|)|(s|)ftp)://")"

-  #cat "$DLLIST"
   #LCK_RM
   #exit
 }

 display_downloadfile(){
   if [ -n "$VERBOSE_OUTPUT" ]; then
-    cat "$DLLIST"
+    echo "$my_dllist"
   else
     DISPLAY_SORT_OPTIONS=(-k 1,1)
     # Sort output after package download size (decreasing):
@@ -538,7 +510,7 @@ if [ "$option" == "install" ]; then
   [ -t 1 ] && [ -n "$REMOVE_WORKING_MESSAGE" ] && tput cuu 1 && tput el && tput cuu 1
   # Test /tmp/apt-fast.list file exists and not just the apt-fast comment line.
   # Then download all files from the list.
-  if [ -f "$DLLIST" ] && [ "$(wc -l "$DLLIST" | cut -d' ' -f1)" -gt 1 ] && [ ! "$DOWNLOADBEFORE" ]; then
+  if [ ! "$DOWNLOADBEFORE" ]; then
     display_downloadfile
     msg
     msg "Do you want to download the packages? [Y/n] " "question"
@@ -560,7 +532,7 @@ if [ "$option" == "install" ]; then
     msg
     # Continue if answer was right or DOWNLOADBEFORE is enabled.
     if ((result)); then
-      if [ -s "$DLLIST" ]; then
+      if true; then
         # Test if apt-fast directory is present where we put packages.
         if [ ! -d "$DLDIR" ]; then
           mkdir -p -- "$DLDIR"
@@ -592,7 +564,7 @@ elif [ "$option" == "clean" ]; then
   "${_APTMGR}" "${APT_SCRIPT_WARNING[@]}" "$@" && {
     find "$DLDIR" -maxdepth 1 -type f -delete
     CLEANUP_STATE="$?"
-    [ -f "$DLLIST" ] && rm -f -- "$DLLIST"* || true
+    true
   }

 elif [ "$option" == "download" ]; then
@@ -604,7 +576,7 @@ elif [ "$option" == "download" ]; then

   [ -t 1 ] && [ -n "$REMOVE_WORKING_MESSAGE" ] && tput cuu 1 && tput el && tput cuu 1

-  if [ -f "$DLLIST" ] && [ "$(wc -l "$DLLIST" | cut -d' ' -f1)" -gt 1 ]; then
+  if true; then
     display_downloadfile
     eval "${_DOWNLOADER}"
   fi
@@ -622,7 +594,7 @@ elif [ "$option" == "source" ]; then

   [ -t 1 ] && [ -n "$REMOVE_WORKING_MESSAGE" ] && tput cuu 1 && tput el && tput cuu 1

-  if [ -f "$DLLIST" ] && [ "$(wc -l "$DLLIST" | cut -d' ' -f1)" -gt 1 ]; then
+  if true; then
     display_downloadfile
     eval "${_DOWNLOADER}"
   fi
```

**Comment by Lasall** [comment]
I close this issue at it was caused by the antivirus program. Please reopen if there are any more evidence of the bug being independent of the antivirus.

**Comment by chrstphrchvz** [comment]
Not all of the symptoms were due to the realtime antivirus. `apt-fast` still has poor performance on WSL compared to "real" Linux; the realtime antivirus just made things noticeably worse.

The only official mention I've found for the underlying issue is https://www.reddit.com/r/bashonubuntuonwindows/comments/69zez8/does_tmpfs_really_work/ (but maybe I can open an issue at https://github.com/Microsoft/WSL/issues for tracking purposes):
> **benhelioz [WSL Developer]** May 8 2017
>You are correct, tmpfs is currently emulated using disk. It's on our backlog to implement it for real.

So the next time to check if this has improved probably won't be until the next major Windows 10 update is out (2019).
